### PR TITLE
rework/networking

### PR DESCRIPTION
**Pull request by Lucaskyy**
followup of #102 

reworked some of the networking, haven't tested it yet though.
some of the patches are from Minestom, possibly adding more.

added support for KQueue
made the packet buffer larger (going to add a config option for that)

and some other things :)

### CODE REVIEW

**Comment by Lucaskyy** [comment]
@hpfxd Does this solve it?

**Review by CyberFlameGO — APPROVED** [review]

**Comment by Lucaskyy** [comment]
FYI: This allows you to manually select a type of Epoll, KQueue, and NIO. If none are selected or the selected type is not supported it will utilize fall-through to select the one that's compatible with your system :)

Never knew fall-through could be handy, haha

**Review by hpfxd — APPROVED** [review]

**Comment by Lucaskyy** [comment]
1 more commit coming soon, if gpg-agent decides to work lul

**Comment by Lucaskyy** [comment]
there we go, does that look good to you now? @hpfxd

**Comment by Lucaskyy** [comment]
it checks if `use-native-transport` is true and the server is running on Linux, if so just use Epoll if not use the `transport-to-use` variable

**Comment by hpfxd** [comment]
If `use-native-transport` is true and the server is on a KQueue-supported platform, it should probably use it, not just for Epoll

**Comment by Lucaskyy** [comment]
@hpfxd Done. I'll merge it now then :)

**Comment by CyberFlameGO** [comment]
> 1 more commit coming soon, if gpg-agent decides to work lul, it's taking forever

ayyy finally someone else who signs their commits

File: NachoSpigot-Server/src/main/java/org/bukkit/craftbukkit/Main.java [hunk 1]
```diff
@@ -20,18 +20,6 @@
...
-//        ---------------------------------------------------------------
-//        if (!SystemUtils.isJavaVersionAtLeast(JavaVersion.JAVA_1_8)) {
-//            System.err.println("TacoSpigot requires java 8");
-//            System.err.println("Oracle dropped all support for java " + SystemUtils.JAVA_VERSION);
-//            System.err.println("Please update to use TacoSpigot and the numerous bug-fixes, performance improvements, and security fixes");
-//            System.err.println("Shutting down");
-//            System.exit(1);
-//        }
         try {
             if(!SystemUtils.isJavaVersionAtLeast(JavaVersion.JAVA_15)) {
```
Comment (crafter23456) [inline NachoSpigot-Server/src/main/java/org/bukkit/craftbukkit/Main.java:24]: Should it be changed to java 16 because protocollib works fine?

Comment (Lucaskyy) [reply NachoSpigot-Server/src/main/java/org/bukkit/craftbukkit/Main.java:24]: We *could*, yes. But Java 15 is recommended either way.

**Comment by bob7l** [comment]
Linux will dynamically adjust the SO_SNDBUF, so setting it shouldn't be needed (Or may even cause problems?). Might only be useful for OS's like Windows.

**Comment by Lucaskyy** [comment]
Oh, didn't know about that. Should I only do that on Windows then? Since as far as I know, SO_SNDBUF tells Netty how big the buffer should be, which in turn allows for more packets to be sent each tick. A bigger Netty buffer should allow for higher play counts, right? At least, that's what I got told.

**Comment by bob7l** [comment]
> 
> 
> Oh, didn't know about that. Should I only do that on Windows then? Since as far as I know, SO_SNDBUF tells Netty how big the buffer should be, which in turn allows for more packets to be sent each tick. A bigger Netty buffer should allow for higher play counts, right? At least, that's what I got told.

There's a min, default, and max size to that buffer on Linux. The max size is gigantic. It'll dynamically adjust it when it needs to (And if there's enough memory to do so...). Windows on the other hand is probably just a static value.

**Comment by Lucaskyy** [comment]
Ah okay. So then, I guess that SO_SNDBUF is not the fix for the Netty buffer lag?

**Comment by bob7l** [comment]
> 
> 
> Ah okay. So then, I guess that SO_SNDBUF is not the fix for the Netty buffer lag?

I doubt it'll effect it. The one confirmed thing this change WILL do however is increase memory usage

**Comment by Lucaskyy** [comment]
Ah okay, thanks for telling me!

**Pull request merged**